Allow `ID_Dedicated`/`null` for `Npc.UserId` Parameter. Additional functionality: - `Npc.LookAt` - `Npc.ShootWeapon` - `Npc.SetAimDownSight` - `Player.UnloadWeapon` - `Player.ToggleWeaponFlashlight`

creepycats opened the pull request:
Use `ID_Dedicated` or `null` to ensure an NPC is VSR Compliant (Uses Unverified/DedicatedHost instances, which will not take up slots or show on the Player List)

Added Multiple Functions for NPC characters to Look at Points, and use a Held Firearm (Shooting, ADS, Flashlight, Reloading)

Had to remake this cause I fricked up branches 😅

creepycats commented:
Cedric on his way to demolish my PR:

NotIntense commented:
Using `ID_Dedicated` for me, as shown in the provided screenshots, doesn't prevent the NPC from showing in the player list/appearing in player count.
![image](https://github.com/Exiled-Team/EXILED/assets/76762610/9f0ba0b3-a50a-4a29-8436-38c6b98e04ad)
![image](https://github.com/Exiled-Team/EXILED/assets/76762610/0fa425e3-f84b-469e-a514-f208f4e9ce89)

VALERA771 commented:
> Using `ID_Dedicated` for me, as shown in the provided screenshots, doesn't prevent the NPC from showing in the player list/appearing in player count.
> 
> ![image](https://github.com/Exiled-Team/EXILED/assets/76762610/9f0ba0b3-a50a-4a29-8436-38c6b98e04ad)
> 
> ![image](https://github.com/Exiled-Team/EXILED/assets/76762610/0fa425e3-f84b-469e-a514-f208f4e9ce89)
> 
> 

Just changing id doesn't change anything if you aren't using this PR

creepycats commented:
Yeah, with the PR, it should set the InstanceMode to DedicatedServer just like my previous fix functions for plugins did, which did appear to work properly.

Also this was literally one of if not the final solutions we had made when looking a few months ago to solve the issue.

creepycats commented:
Update: I DID miss something Lmfao, you cant set `npc.ReferenceHub.authManager.UserId` with the Dedicated Server like I did. You have to instead use `npc.ReferenceHub.authManager.SyncedUserId`. Ill commit the change real quick.

NotIntense commented:
> > Using `ID_Dedicated` for me, as shown in the provided screenshots, doesn't prevent the NPC from showing in the player list/appearing in player count.
> > ![image](https://private-user-images.githubusercontent.com/76762610/285501306-9f0ba0b3-a50a-4a29-8436-38c6b98e04ad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDA4Nzc2NjYsIm5iZiI6MTcwMDg3NzM2NiwicGF0aCI6Ii83Njc2MjYxMC8yODU1MDEzMDYtOWYwYmEwYjMtYTUwYS00YTI5LTg0MzYtMzhjNmI5OGUwNGFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI1VDAxNTYwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZkMzkwMmMwZDViMDc0ODJkMmRlMWZhMGU1Njc2OGQxNmM4ZGQzN2NiYzMwZjI3MzNmYjg3NDFkNzZlMGFmZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Dr3pFGCY3lRER3z711WdjwI2ookwZe6DdXsuYUb6H40)
> > ![image](https://private-user-images.githubusercontent.com/76762610/285501323-0fa425e3-f84b-469e-a514-f208f4e9ce89.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDA4Nzc2NjYsIm5iZiI6MTcwMDg3NzM2NiwicGF0aCI6Ii83Njc2MjYxMC8yODU1MDEzMjMtMGZhNDI1ZTMtZjg0Yi00NjllLWE1MTQtZjIwOGY0ZTljZTg5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI1VDAxNTYwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEyNjVkMGVkN2UzNTUzZmUzNDg5MjllMDY2NzdmM2Y3Yjg5Yzk5OTVmYWY4ZmZlZmEwYTdiYzA1ZDBkMGJmN2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.U0zm3Lsczr0Q4Uz2bWEQlFJOveTGkWt9mnZBwp-t_Ys)
> 
> Just changing id doesn't change anything if you aren't using this PR

Dumbass, I was :trollface: :

creepycats commented:
I've also added support for `null` alongside `ID_Dedicated`. This gives you the option of plain Unverified players and Dedicated Server players. Issue should be resolved.
![image](https://github.com/Exiled-Team/EXILED/assets/68137105/c0dd1af4-9f5d-4d30-82c9-cc00ae8f8557)

VALERA771 commented on Exiled.API/Features/Npc.cs:
```diff
@@ -196,5 +213,125 @@ public void Destroy()
+
+            Physics.Raycast(CameraTransform.position, CameraTransform.forward, out RaycastHit hit, 100f, StandardHitregBase.HitregMask);
+
+            if (hit.transform && hit.transform.TryGetComponentInParent(out NetworkIdentity networkIdentity) && networkIdentity)
```
I think instead of `NetworkIdentety` here should be `IDestructable`. Iirc base game shooting logic works like this

VALERA771 commented on Exiled.API/Features/Npc.cs:
```diff
@@ -196,5 +213,125 @@ public void Destroy()
+        /// Makes the NPC Shoot at the specified location if holding a <see cref="Firearm"></see>.
+        /// </summary>
+        /// <param name="targetPos">The position to shoot towards.</param>
+        public void Shoot(Vector3? targetPos = null)
```
Can be converted to book and return false if player is not holding a firearm or not IFpcRole etc

creepycats commented:
Did my best to get it working before I had to leave. Noticed I also overrode the shooting position when hitting players. Anyways, I simplified the if returns into one function, and use an IDestructible instead of NetworkIdentity.

creepycats commented:
Gonna change more things and do comment stuff.

creepycats commented:
Should be all done. Might need to fix a bug or two if it comes up, but this should work.

- Function Names are Clearer
- They return a bool depending on if the requests are received or if the player is not able to send the request.
- Added Documentation for the bool returns
- I THINK I shortened the two if->return statements into one for each one 
(`if (RoleManager.CurrentRole is not IFpcRole fpc || ReferenceHub.inventory._curInstance is not Firearm firearm)`)
- `NetworkIdentity` replaced with `IDestructible`, using Game's own weapon logic

creepycats commented:
Fixed Documentation with past commits, and also removed the ReloadWeapon function from NPCs since Player apparently has it already. I might move everything except for the ShootWeapon function in the Weapon stuff to Player as well.

creepycats commented:
Quickly moved my weapon functions to `Player.cs` from `NPC.cs`. This is NOT tested and I will be modifying them to fit the current function for Reloading. Will look for review afterwards.

creepycats commented:
Should be good to go.

- Consistency with `ReloadWeapon` for others
- Moved `LookAt`, `ShootWeapon` and `SetAimDownSight` back to NPC since they only seem to work properly there.
- Rewrote `LookAt` since it for some reason didn't work with players or at all using the shortcut method, so I use smth else
- Removed `Lerp` parameter for `LookAt`
- Converted InventorySystem `Firearm` to be based on Exiled's `Firearm`
- Few Slight Fixes

All was tested privately. VSR-options for NPCs (null/`ID_Dedicated`) works perfectly, and all Player tied functions work on both the client and server. Any function that didnt was moved to NPC specifically for their control (AKA Looking around, Shooting, and ADSing)

creepycats commented:
Made two small changes to Documentation, sorry it took so long i got eepy
